spikeinterface/curation/remove_duplicated_spikes.py

```diff
@@ -0,0 +1,159 @@
+		if end_frame == None:
+			end_frame = spike_train[-1]
+
+		start = np.searchsorted(spike_train, start_frame, side="left")
+		end   = np.searchsorted(spike_train, end_frame, side="right")
+
+		return spike_train[start : end]
+
+
+remove_duplicated_spikes = define_function_from_class(source_class=RemoveDuplicatedSpikesSorting, name="remove_duplicated_spikes")
+
+
+def find_duplicated_spikes_random(spike_train: np.ndarray, censored_period: int, seed: Optional[int] = 186472189) -> np.ndarray:
+	"""
+	Finds the indices where there a spike in considered a duplicate.
+	When two spikes are closer together than the censored period,
+	one of them is randomly taken out.
+	Randomness allows to make sure no bias is generated by always
+	keeping the first or last spike. However, it may not always remove
+	the same number of spikes each time it is called.
+
+	Parameters
+	----------
+	spike_train: np.ndarray
+		The spike train on which to look for duplicated spikes.
+	censored_period: int
+		The censored period for duplicates (in sample time).
+	seed: int
+		The random seed for taking out spikes.
+
+	Returns
+	-------
+	indices_of_duplicates: np.ndarray
+		The indices of spikes considered to be duplicates.
+	"""
+
+	rand_state = np.random.get_state()	# Need to store the old state to not seed globally.
+	np.random.seed(seed)				# Need to seed to have the same result for parallelization.
+
+	indices_of_duplicates = np.where(np.diff(spike_train) <= censored_period)[0]
+	indices_of_duplicates = np.unique(np.concatenate((indices_of_duplicates, indices_of_duplicates + 1)))
+	mask = np.ones(len(indices_of_duplicates), dtype=bool)
+
+	# While there are still some duplicated spikes detected, remove one of them at random.
+	while not np.all(np.diff(spike_train[indices_of_duplicates][mask]) > censored_period):
+		index = np.random.randint(low=0, high=len(mask))
+		mask[index] = False
+
+	np.random.set_state(rand_state)
+
+	return indices_of_duplicates[~mask]
+
+if HAVE_NUMBA:
+	@numba.jit((numba.int64[::1], numba.int32), nopython=True, nogil=True, cache=True, parallel=True)
+	def find_duplicated_spikes_keep_first(spike_train, censored_period):
+		indices_of_duplicates = numba.typed.List()
+		N = len(spike_train)
+
+		for i in range(N-1):
+			if i in indices_of_duplicates:
+				continue
+
+			for j in range(i+1, N):
+				if spike_train[j] - spike_train[i] > censored_period:
+					break
+				indices_of_duplicates.append(j)
+
+		return np.asarray(indices_of_duplicates)
+
+	@numba.jit((numba.int64[::1], numba.int32), nopython=True, nogil=True, cache=True, parallel=True)
+	def find_duplicated_spikes_keep_last(spike_train, censored_period):
+		indices_of_duplicates = numba.typed.List()
+		N = len(spike_train)
+
+		for i in range(N-1, 0, -1):
+			if i in indices_of_duplicates:
+				continue
+
+			for j in range(i-1, -1, -1):
+				if spike_train[i] - spike_train[j] > censored_period:
+					break
+				indices_of_duplicates.append(j)
+
+		return np.asarray(indices_of_duplicates)
```

Comment:
could this be the same function? you can just define the iterator based on the method.

Reply:
Done.